### Resources
- http://joelonsoftware.com/articles/Unicode.html 
- https://www.youtube.com/watch?v=MijmeoH9LT4
- https://www.youtube.com/watch?v=sgHbC6udIqc

In [1]:
import pandas as pd
train=pd.read_csv('Train.csv')
sample=pd.read_csv('sample_submission.csv', encoding='utf-8', index_col= 'objectId')

What lives inside the train datasets 'product' column?

In [2]:
one_product=train['product'][0]
print one_product
type(one_product)

túró rudi


str

What if we try to decode it? 

In [3]:
one_product.decode('utf-8')

u't\xfar\xf3 rudi'

Althoug the notebook output is a bit crusty, the utf-8 decosed unicode test-representation is the was to go. Here the "print" function wil help the reader, and the 'type' shows, that the variable is unicode.

In [4]:
unicode_one_product= one_product.decode('utf-8')
print unicode_one_product
print type(unicode_one_product)

túró rudi
<type 'unicode'>


In [5]:
one_product == unicode_one_product

/home/datacarpenter/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:1: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
  if __name__ == '__main__':


False

They look similar, but because the different type, the equalit returns "False". it would be better to throw, an error.

In [6]:
unicode_one_product.decode('utf-8')

UnicodeEncodeError: 'ascii' codec can't encode character u'\xfa' in position 1: ordinal not in range(128)

This erro shows us the main concerne to keep in mind: Python 2.7 tries to implicitly (secterly)  encode everything into ASCII.  'sys.getdefaultencoding()' shows us why

In [7]:
import sys
sys.getdefaultencoding()

'ascii'

In [8]:
simple_string='alma'
u_string=u'alma'
print type(simple_string)
print type(u_string)
simple_string == u_string

<type 'str'>
<type 'unicode'>


True

altohug they have a differen type 'alma' ans u'alma' happens to be equal. This is considered as a FEATURE by the Py 2.7 developers, but if you happen to have only-unicode chars in your text, it becomes a BUG. 

In [9]:
all_train_products= train['product'].value_counts().index.tolist()
type(all_train_products[0])


str

Lets check out our other source file, the sample submission. It contains unicode characters. 

In [10]:
all_sample_products= sample.columns.tolist()
type(all_sample_products[0])

unicode

Try to compare the two lists: how many are different?

In [11]:
len([x for x in all_sample_products if not x in all_train_products])

/home/datacarpenter/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:1: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
  if __name__ == '__main__':


142

142 products which are missing?  Try it again with proper types. (unicode)

In [12]:
all_train_producst_unicode= [x.decode('utf-8') for x in all_train_products]
len([x for x in all_sample_products if not x in all_train_producst_unicode])

0

Thats better! now we should create a new columns with unicode values in the train table.  

In [13]:
def create_unicode(text):
    return text.decode('utf-8')

train['product_uni']= train['product'].apply(create_unicode)

which equals to

In [14]:
train['product_uni']= train['product'].apply(lambda x: x.decode('utf-8'))

In [15]:
train['product_uni']

0           túró rudi
1              zeller
2               liszt
3             kolbász
4                alma
5                 tea
6              kenyér
7            szalvéta
8               üdítő
9            tusfürdő
10           gyümölcs
11             zsemle
12          ásványvíz
13             tejföl
14         barnacukor
15             öblítő
16         darált hús
17            fogkrém
18              málna
19                tej
20             öblítő
21             kenyér
22                tej
23                tea
24          fokhagyma
25          felvágott
26                tej
27            vajkrém
28             dohány
29                vaj
             ...     
22266             tej
22267    zsemlemorzsa
22268      csirkemell
22269         mosópor
22270          pékáru
22271         paprika
22272           kefír
22273          uborka
22274       zabpehely
22275          kenyér
22276         narancs
22277          tejföl
22278         kalcium
22279         szalámi
22280     